packages/rax/src/markupChecksum.js

```diff
@@ -0,0 +1,53 @@
+const MOD = 65521;
+const CHECKSUM_ATTR_NAME = 'data-rax-checksum';
+const TAG_END = /\/?>/;
+
+// adler32 is not cryptographically strong, and is only used to sanity check that
+// markup generated on the server matches the markup generated on the client.
+// This implementation (a modified version of the SheetJS version) has been optimized
+// for our use case, at the expense of conforming to the adler32 specification
+// for non-ascii inputs.
+export function adler32(data) {
+  let a = 1;
+  let b = 0;
+  let i = 0;
+  let l = data.length;
+  let m = l & ~0x3;
+  while (i < m) {
+    let n = Math.min(i + 4096, m);
+    for (; i < n; i += 4) {
+      b += (a += data.charCodeAt(i)) + (a += data.charCodeAt(i + 1)) + (a += data.charCodeAt(i + 2)) + (a += data.charCodeAt(i + 3));
+    }
+    a %= MOD;
+    b %= MOD;
+  }
+  for (; i < l; i++) {
+    b += a += data.charCodeAt(i);
+  }
+  a %= MOD;
+  b %= MOD;
+  return a | b << 16;
+}
+
+
+/**
+  * add checksum to markup
+*/
+export function addChecksumToMarkup(markup) {
```

Comment:
rax can return a node list, if add checksum to serializer, i have to add checksum to not only one node. maybe i can wrap them and add checksum to wrapper. in browser, check the wrapper instead of check every node in the wrapper.